tests/link/test_numba.py

```diff
@@ -2893,3 +2894,144 @@ def test_random_Generator():
+    et0 = aet.lscalar("e_t0")
+    it0 = aet.lscalar("i_t0")
+    logp_c = aet.scalar("logp_c")
+    logp_d = aet.scalar("logp_d")
+    # non_sequences
+    beta = aet.scalar("beta")
+    gamma = aet.scalar("gamma")
+    delta = aet.scalar("delta")
+
+    def seir_one_step(ct0, dt0, st0, et0, it0, logp_c, logp_d, beta, gamma, delta):
+        bt0 = st0 * beta
+        bt0 = bt0.astype(st0.dtype)
+
+        logp_c1 = binom_log_prob(et0, gamma, ct0).astype(logp_c.dtype)
+        logp_d1 = binom_log_prob(it0, delta, dt0).astype(logp_d.dtype)
+
+        st1 = st0 - bt0
+        et1 = et0 + bt0 - ct0
+        it1 = it0 + ct0 - dt0
+        return st1, et1, it1, logp_c1, logp_d1
+
+    (st, et, it, logp_c_all, logp_d_all), _ = scan(
+        fn=seir_one_step,
+        sequences=[aet_C, aet_D],
+        outputs_info=[st0, et0, it0, logp_c, logp_d],
+        non_sequences=[beta, gamma, delta],
+    )
+    st.name = "S_t"
+    et.name = "E_t"
+    it.name = "I_t"
+    logp_c_all.name = "C_t_logp"
+    logp_d_all.name = "D_t_logp"
+
+    out_fg = FunctionGraph(
+        [aet_C, aet_D, st0, et0, it0, logp_c, logp_d, beta, gamma, delta],
+        [st, et, it, logp_c_all, logp_d_all],
+    )
+
+    s0, e0, i0 = 100, 50, 25
+    logp_c0 = np.array(0.0, dtype=config.floatX)
+    logp_d0 = np.array(0.0, dtype=config.floatX)
+    beta_val, gamma_val, delta_val = [
+        np.array(val, dtype=config.floatX) for val in [0.277792, 0.135330, 0.108753]
+    ]
+    C = np.array([3, 5, 8, 13, 21, 26, 10, 3], dtype=np.int32)
+    D = np.array([1, 2, 3, 7, 9, 11, 5, 1], dtype=np.int32)
+
+    test_input_vals = [
+        C,
+        D,
+        s0,
+        e0,
+        i0,
+        logp_c0,
+        logp_d0,
+        beta_val,
+        gamma_val,
+        delta_val,
+    ]
+    compare_numba_and_py(out_fg, test_input_vals)
+
+
+@config.change_flags(compute_test_value="raise")
+def test_scan_tap_output():
+
+    a_aet = aet.scalar("a")
+    a_aet.tag.test_value = 10.0
+
+    b_aet = aet.arange(10).astype(config.floatX)
+    b_aet.name = "b"
+
+    c_aet = aet.arange(20, 30, dtype=config.floatX)
+    c_aet.name = "c"
+
+    def input_step_fn(b, c, x_tm1, y_tm1, y_tm3, a):
+        x_tm1.name = "x_tm1"
+        y_tm1.name = "y_tm1"
+        y_tm3.name = "y_tm3"
+        y_t = (y_tm1 + y_tm3) * a + b
+        z_t = y_t * c
+        x_t = x_tm1 + 1
+        x_t.name = "x_t"
+        y_t.name = "y_t"
+        return x_t, y_t, z_t
+
+    scan_res, _ = scan(
+        fn=input_step_fn,
+        sequences=[b_aet, c_aet],
+        outputs_info=[
+            {
+                "initial": aet.as_tensor_variable(0.0, dtype=config.floatX),
+                "taps": [-1],
+            },
+            {
+                "initial": aet.as_tensor_variable(
+                    np.r_[-1.0, 1.3, 0.0].astype(config.floatX)
+                ),
+                "taps": [-1, -3],
+            },
+            None,
+        ],
+        non_sequences=[a_aet],
+        # n_steps=10,
```

Comment:
One of the current issue with this particular implementation is that when we provide the `n_steps` explicity , it is provided to the Numba function as an 0-d array unlike when `scan` derives it on its own (the case in which it is provided as an `int`). Hence during initial compilation's type-checking the code that follows the `n_steps` in the function produced i.e. ```for i in range(n_steps)``` fails. Since Numba doesn't have an inbuilt type for range with 0-d array.

Reply:
Sounds like `.item` is needed when it's a 0d array, and not otherwise.  There's some existing code that conditions on these situations.